### ISSUE TITLE
`request.referer` and session variable obstacles

### ISSUE THREAD

**Issue by jonathanhefner** [kaspth/action_controller-stashed_redirects#3]
I have been thinking about an API like this, and I am very excited to see someone else pursue it! :smiley:

I wanted to share a few obstacles that I encountered while trying to use `request.referer` and / or session variables:

* Redirects preserve the original `Referer` header.  For example:

    1\. Visit `/posts/1` while signed out.
    2\. Click `/posts/1/edit` link.  `Referer` will be `/posts/1`.
    3\. Require authentication: redirect from `/posts/1/edit` to `/sign_in`.
    4\. Serve `/sign_in`.  `Referer` is `/posts/1`, *not* `/posts/1/edit`.

    So if `stash_redirect_for` stashes `request.referer` in the above example, `redirect_from_stashed` will redirect to `/posts/1`, not `/posts/1/edit`.

* Some browsers (e.g. Firefox) have privacy settings to spoof the `Referer` header.  In Firefox, it appears that the spoofed value is the page itself.  So, in the previous example, `Referer` would instead be `/sign_in`.

* As shared state, session variables can be a problem when using multiple browser tabs.  For example:

    1\. Visit `/posts` while signed out.
    2\. Open `/posts/1/edit`, `/posts/2/edit`, etc links in new tabs.
    3\. Each tab is redirected to `/sign_in`.  The final tab "wins" at setting the session variable.
    4\. Assume forms are autofilled by a password manager.  User submits each tab's form because it is easier than reopening each tab.
    5\. The first tab is redirected to the final tab's intended location.  If the session variable is then discarded, `redirect_from_stashed` will raise an `ActionController::StashedRedirects::MissingRedirectError` for all other tabs.

* Session variables can erroneously "follow" the user.  For example:

    1\. Visit `/posts/1` while signed out.
    2\. Click `/posts/1/edit` link.
    3\. Require authentication: invoke `stash_redirect_for(:sign_in, from: request.fullpath)` *before* the redirect (instead of `::stash_redirect_for` after the redirect), then redirect to `/sign_in`.
    4\. Navigate to a page that doesn't require authentication, e.g. `/`.
    5\. Click `/sign_in` link in the page navbar.  Sign in.
    6\. `redirect_from_stashed` will redirect to `/posts/1/edit`, despite starting a new navigation sequence, because nothing triggered the session variable to change.

My conclusion has been to implement a [query param based API](https://github.com/jonathanhefner/rails/commit/0e277f1c559047b31e0982f7436a935adcdece43), but you might come up with something better!  I was planning to submit a PR to Rails, but I will hold off if you think there is a better approach.

**Comment by kaspth** [comment]
@jonathanhefner hey, thanks for checking this out!

> Redirects preserve the original Referer header.

I didn't know this and now I'm confused. How does `redirect_back_or_to` work then because it relies on `request.referer`, would you happen to know?

> As shared state, session variables can be a problem when using multiple browser tabs

I just tried to check precedent through Devise and it looks like they use session variables too. I think we're mitigating this somewhat by requiring a purpose be specified.

> My conclusion has been to implement a [query param based API](https://github.com/jonathanhefner/rails/commit/0e277f1c559047b31e0982f7436a935adcdece43), but you might come up with something better!

Yeah, we do support that `stash_redirect_for :sign_in, from: :param`, which fetches from `params[:redirect_url]` and takes precedence over `referer`. Though we could potentially snip the `referer` support.

**Comment by kaspth** [comment]
> Regardless, if such a feature is added to Rails

I left core back in march and have no intention of contributing to Rails again, I'll be doing these kinds of support gems though.

I'll check back later on the other stuff, but I won't be submitting this to Rails.

**Comment by jonathanhefner** [comment]
Oh, I misunderstood!  I thought the intention was to eventually pitch this feature to Rails core (after refining it here).  Okay, I will submit my PR then, but please let me know if you have any suggestions!

**Comment by kaspth** [comment]
> but please let me know if you have any suggestions!

Yeah, I'm not still interested in having anything to do with Rails core.
